text/0000-quick-debug-macro.md

```diff
@@ -0,0 +1,177 @@
+
+The motivation to add this new `dbg!` macro is two-fold.
+
+## For aspiring rustaceans
+
+One of the often asked questions is how to print out variables to the terminal.
+Delaying the need to explain formatting arguments in the statement `println!("{:?}", expr);` can help aspiring rustaceans to quickly learn the language. With `dbg!(expr);` there is no longer such need, which can be delayed until the developer actually cares about the format of the output and not just what value the expression evaluates to.
+
+## For experienced developers
+
+By using `dbg!(expr);`, the burden of a common papercut: writing `println!("{:?}", expr);` every time you want to see the evaluted-to value of an expression, can be significantly reduced. The developer no longer has to remember the formatting args and has to type significantly less (12 characters to be exact).
+
+The shortness is also beneficial when asking `rustbot` on #rust@irc.mozilla.org to evaluate and print an expression.
+
+To increase the utility of the macro, it acts as a pass-through function on the expression by simply printing it and then yielding it. On release builds, the macro is the identity function - thus, the macro can be used in release builds while hurting performance while also helping to debug the program.
+
+## Why not use the `log` crate?
+
+While the `log` crate offers a lot of utility, it first has to be used with `extern crate log;`. A logger then has to be set up before expressions can be logged. It is therefore not suitable for introducing newcommers to the language.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## On debug builds:
+
+First, some preliminaries:
+
+```rust
+#[derive(Debug)] // The type of expr in dbg!(expr) must be Debug.
+struct Point {
+    x: usize,
+    y: usize,
+}
+```
+
+With the following example (which most newcomers will benefit from):
+
+```rust
+fn main() {
+    dbg!(Point {
+        x: 1,
+        y: 2,
+    });
+
+    let p = Point {
+        x: 4,
+        y: 5,
+    };
+    dbg!(p);
+}
+```
+
+The program will print the points to `STDERR` as:
+
+```
+[DEBUGGING, src/main.rs:1:4]:
+=> Point{x: 1, y: 2,} = Point {
+    x: 1,
+    y: 2
+}
+[DEBUGGING, src/main.rs:7:4]:
+=> p = Point {
+    x: 4,
+    y: 5
+}
+```
+
+Here, `7:4` is the line and the column.
+
+You may also save the debugged value to a variable or use it in an expression
+since the debugging is pass-through:
+
+```rust
+fn main() {
+    let x = dbg!(1 + 2);
+    let y = dbg!(x + 1) + dbg!(3);
+    dbg!(y);
+}
+```
+
+This prints the following to `STDERR`:
+
+```
+[DEBUGGING, src/main.rs:1:12]:
+=> 1 + 2 = 3
+[DEBUGGING, src/main.rs:2:12]:
+=> x + 1 = 4
+[DEBUGGING, src/main.rs:2:26]:
+=> 3 = 3
+[DEBUGGING, src/main.rs:3:4]:
+=> y = 7
+```
+
+This way of using the macro will mostly benefit existing Rust programmers.
+
+## On release builds:
+
+The same examples above will print nothing to `STDERR` and will instead simply
```

Comment:
Why? `dbg!()` calls don't seem like something that you'd want to keep around other than when you're looking into a specific bug.

Reply:
I think looking into a specific bug is the most likely scenario. However...
There might be some fix you made, but you are not 100% sure, and you'd still like to include it into release builds. With the macro having no effect on `release`, this is possible.

Does this have any downsides?

Reply:
Spewing a bunch of stuff to stderr when a downstream crate isn't building with `--release` is a downside.

Asserts are a more common way of handling those kinds of checks IME.

Reply:
I buy the argument regarding `assert` for at least libraries. However, it might have some value for programs?